feeder: policyMatcher: Refactoring SecurityPolices functions and adding matchPatterns logic

geyslan opened the pull request:
This refactors the `UpdateSecurityPolicies()` and  `UpdateHostSecurityPolicies()` providing the same logic
with less code through helper functions, reducing checking and execution.

This also includes logic for `matchPatterns` and adds the `ksp-ubuntu-1-file-pattern-block.yaml` security policy example.

As an additional implementation, this proposes changes to the `KubeArmor/types/types.go:MatchPolicy` structure by adding a `Regexp` pointer field to it.

Fixes #143

Commit:
feeder: policyMatcher: UpdateSecurityPolicies() Refactoring

This refactors the `UpdateSecurityPolicies()` providing the same logic
with less code through helper functions, reducing checking and execution.

Commit:
feeder: policyMatcher: Add matchPatterns logic

geyslan commented on KubeArmor/feeder/policyMatcher.go:
```diff
@@ -12,450 +13,348 @@ import (
+		Operation:  "Process",
+		Resource:   path.Path,
+		Action:     path.Action,
+	}
```
These are just helper functions. It could be managed through interfaces instead of helpers, but as a first proposal I preferred not to touch types.go concerning that. Another additional discussion would be related to the use of embedded structure/interface types. I noticed that some have similar structure.

geyslan replied:
@nam-jaehyun this can be changed to:

```golang
func newMatchPolicy(policyName, src, op string, mp interface{}) tp.MatchPolicy {
	match := tp.MatchPolicy{
		PolicyName: policyName,
		Source:     src,
		Operation:  op,
	}

	if ppt, ok := mp.(tp.ProcessPathType); ok {
		match.Severity = strconv.Itoa(ppt.Severity)
		match.Tags = ppt.Tags
		match.Message = ppt.Message
		match.Resource = ppt.Path
		match.Action = ppt.Action
	} else if pdt, ok := mp.(tp.ProcessDirectoryType); ok {
		match.Severity = strconv.Itoa(pdt.Severity)
		match.Tags = pdt.Tags
		match.Message = pdt.Message
		match.Resource = pdt.Directory
		match.Action = pdt.Action
	} else if fpt, ok := mp.(tp.FilePathType); ok {
		match.Severity = strconv.Itoa(fpt.Severity)
		match.Tags = fpt.Tags
		match.Message = fpt.Message
		match.Resource = fpt.Path
		match.Action = fpt.Action
	} else if fdt, ok := mp.(tp.FileDirectoryType); ok {
		match.Severity = strconv.Itoa(fdt.Severity)
		match.Tags = fdt.Tags
		match.Message = fdt.Message
		match.Resource = fdt.Directory
		match.Action = fdt.Action
	} else if npt, ok := mp.(tp.NetworkProtocolType); ok {
		match.Severity = strconv.Itoa(npt.Severity)
		match.Tags = npt.Tags
		match.Message = npt.Message
		match.Resource = getProtocolFromName(npt.Protocol)
		match.Action = npt.Action
	} else if cct, ok := mp.(tp.CapabilitiesCapabilityType); ok {
		match.Severity = strconv.Itoa(cct.Severity)
		match.Tags = cct.Tags
		match.Message = cct.Message
		match.Resource = getCapabilityFromName(cct.Capability)
		match.Action = cct.Action
	} else {
               // type not found
        }

	return match
}
```

It could be smaller but we lack of common structures/interfaces.

geyslan commented on KubeArmor/feeder/policyMatcher.go:
```diff
@@ -12,450 +13,348 @@ import (
-							}
-						}
-					}
+			for _, src := range path.FromSource {
+				if len(src.Path) > 0 {
+					fromSource = src.Path
+				} else if len(src.Directory) > 0 {
+					fromSource = src.Directory
+				} else {
+					continue
```
This is a XOR checking as I've mentioned at https://github.com/accuknox/KubeArmor/pull/152#issuecomment-862447832

geyslan commented on KubeArmor/feeder/policyMatcher.go:
```diff
@@ -12,450 +13,348 @@ import (
 
-								case "ICMP", "icmp":
-									match.Resource = "type=SOCK_RAW protocol=1"
+			regexpComp, err := regexp.Compile(patt.Pattern)
+			if err != nil {
+				continue
+			}
```
This is a proposal based on https://github.com/accuknox/KubeArmor/issues/143#issuecomment-862113552 and the part of this PR that would Fix #143. It demanded a change in the types.go, adding Regexp field to `MatchPolicy` what I know isn't the right approach (draft for discussion).

As I've mentioned in the #143, Globbing patterns aren't Regular Expressions _per si_. AppArmor has its proper Globbing syntax and logic expansions which are very similar to bash ones with addition of some sugar. E.g.:

```
/dir/**  - match any file or directory in /dir/ or any directory below /dir/
/dir/**/ - match any directory in /dir/ or any directory below /dir/
```
Related to that we have two paths:
1. Using proper Regular Expressions `regexp.Regexp`
2. Using Globbing without syntax sugar `path.filepath.Glob` and/or `path.filepath.Match`

Following 2, we will have issues and lack of functionality. https://github.com/accuknox/KubeArmor/issues/143#issuecomment-861596803

For Globbing with syntax sugar, we can consider using [glob](https://github.com/gobwas/glob) which also promises better [performance](https://github.com/gobwas/glob#performance) than Regexp.

geyslan replied:
As approved in https://github.com/accuknox/KubeArmor/pull/152#issuecomment-866046860, I presume using regexp.Regexp for now is the right approach. Marking as solved for now.

nam-jaehyun commented on KubeArmor/feeder/policyMatcher.go:
```diff
@@ -12,450 +13,348 @@ import (
-						match.Source = ""
-						match.Operation = "Process"
-						match.Resource = dir.Directory
+func newCapMatchPolicy(policyName, source, resource string, cap tp.CapabilitiesCapabilityType) tp.MatchPolicy {
```
Here, net_raw capability is for networking; thus, the operation was network.
However, there are other capabilities. Please make room for them.

geyslan replied:
Ok. I'm on it. Thanks.

Commit:
feeder: policyMatcher: helper functions merged and minor fixes

Commit:
feeder: policyMatcher: function header comments and minor fix

Commit:
feeder: policyMatcher: UpdateHostSecurityPolicies() Refactoring

This refactors the `UpdateHostSecurityPolicies()` providing the same logic
with less code through helper functions, reducing checking and execution.

Commit:
examples: add file pattern block yaml
